### PR TITLE
Make security-spring respect runtime configuration

### LINKED ISSUE #4905: security-spring does not honour runtime configuration
PR #4787 moved the recommended way to configure installations to `$PORTAL_HOME/portal.properties`. However, the `security/security-spring/` component (or at least the Keycloak integration code within it) bypasses the central logic on where configuration comes from and instead uses `applicationContext-security.xml`.

### PR DESCRIPTION
**Pull request by fedde-s**
In PR #4787, I adjusted the central logic on reading configuration files, as
well as the deployment manual, moving the main way to supply
configuration to $PORTAL_HOME/portal.properties at runtime. Later, I
found out that the security-spring applicationContext uses its own logic to read the same
files, and this logic was broken.

As a quick fix to make configuration of auth settings work again, adjust
security-spring's duplicate logic to behave like GlobalProperties at
least when these settings are actually listed in the runtime
portal.properties.

Turn `authenticate=` into a required JVM system property (`-D` arg) when using runtime configuration, as it's used before `portal.properties` is read.

Fix #4905. The main issue, at least; the remaining maintainability and deployability issue I logged as #4916.

I pointed this pull request to `rc` rather than `master` even though it's a bugfix, because it only really matters after #4787.

# Checks
- [x] Runs on Heroku.
- [x] Follows [7 rules of great commit messages](http://chris.beams.io/posts/git-commit/). For most PRs a single commit should suffice, in some cases multiple topical commits can be useful. During review it is ok to see tiny commits (e.g. Fix reviewer comments), but right before the code gets merged to master or rc branch, any such commits should be squashed since they are useless to the other developers. Definitely avoid [merge commits, use rebase instead.](http://nathanleclaire.com/blog/2014/09/14/dont-be-scared-of-git-rebase/)
- [ ] Follows the [Google Style Guide](https://github.com/google/styleguide).
- [ ] If this is a feature, the PR is to rc. If this is a bug fix, the PR is to master.

### CODE REVIEW

**Comment by fedde-s** [comment]
I'm still testing whether the fix actually works.

**Comment by fedde-s** [comment]
It doesn't seem to, but I'm not familiar enough with Spring to figure this out on my own. Is the syntax I'm trying to use correct? Previously, I believe, the XML file tried to use the `propertyPlaceholderConfigurer` bean to determine the value of `$PORTAL_HOME` as part of a parameter to initialise the `propertyPlaceholderConfigurer` bean. I tried a different approach in commit https://github.com/cBioPortal/cbioportal/pull/4787/commits/39c6c073902d5b5f02b6999f35436ecc9947430c and copied it here, but now I still suspect that [`web.xml`](https://github.com/cBioPortal/cbioportal/blob/39c6c073902d5b5f02b6999f35436ecc9947430c/portal/src/main/webapp/WEB-INF/web.xml#L26) is not setting the right profile based on `$PORTAL_HOME/portal.properties` at runtime.

**Comment by fedde-s** [comment]
I found something really puzzling in `web.xml`. It looks like Maven transforms these lines:
```xml
    <context-param>
        <param-name>spring.profiles.active</param-name>
        <param-value>${dbconnector:jndi},${authenticate}</param-value>
    </context-param>
```
into these:
```xml
    <context-param>
        <param-name>spring.profiles.active</param-name>
        <param-value>${dbconnector:jndi},false</param-value>
    </context-param>
```

Making the authenticate setting non-configurable.

**Comment by fedde-s** [comment]
I think that requires changes to some `pom.xml` file?

**Comment by fedde-s** [comment]
No, that probably wouldn't work. I think `web.xml` cannot access the configuration in `portal.properties` at all, so if we want to make `authenticate` configurable and use it in `web.xml`, it would have to be a Java system property, supplied to the JVM via a `-D` parameter in `CATALINA_OPTS`.

**Comment by fedde-s** [comment]
I can't currently think of a way to set the profile that satisfies all three of these criteria:
- [ ] behaves as before for installations that have authentication enabled in the buildtime properties and do not supply runtime properties,
- [ ] respects the runtime properties if they are supplied and the buildtime properties were left untouched (leaving Maven to copy in `portal.properties.EXAMPLE`), and
- [ ] defaults to `false` if the runtime properties do not mention `authenticate` and the buildtime properties were left untouched (leaving Maven to copy in `portal.properties.EXAMPLE`)

**Comment by fedde-s** [comment]
A solution that basically satisfies two of these, would be removing the `authenticate` property from `portal.properties.EXAMPLE`, and documenting that it should *always* be supplied as `-D` property in `$CATALINA_OPTS` (see https://github.com/cBioPortal/cbioportal/issues/4832). It would probably mean that the server crashes if neither is supplied though.

**Comment by mandawilson** [comment]
@fedde-s We run multiple web apps on the same Tomcat with different authentication methods.  We would have to segregate our web apps into different Tomcats where the authentication method needed was the same for all apps, is that correct?  Or we could just choose to continue to put the authenticate property into our application-specific classpath:portal.properties files, correct?  We would continue to things as we used to, building different wars for each application.

**Comment by fedde-s** [comment]
The solution I proposed would still be a step forward though: everyone else, deploying a separate process for each cBioPortal server (maybe using Docker), can start using runtime configuration that way.

**Comment by fedde-s** [comment]
Unless you guys have an idea on how to access the `authenticate=` property from `portal.properties` to set the profile before `web.xml` starts `applicationContext-security.xml`, the code I pushed now gets runtime configuration to work without breaking pre-existing buildtime Maven filtering.

It does come with the somewhat awkward requirement that the JVM parameter must *always* be supplied if not using the old buildtime configuration setup, as I'm not aware of a syntax to set the default that can interchangeably be parsed by both Maven and Spring.

**Comment by fedde-s** [comment]
@pieterlukasse suggested that not having a default for the `authenticate=` property may actually be a good thing: it can help avoid accidentally publishing sensitive or unlicensed data.

**Comment by fedde-s** [comment]
I tested it with a different Dockerfile, and found out that something still seems to get hard-coded from `portal.properties.EXAMPLE` somewhere. If I don't replace `portal.properties.EXAMPLE` with an empty file, I get this error at runtime:

```
Failed to instantiate [org.springframework.security.saml.key.JKSKeyManager]: Constructor threw exception; nested exception is java.lang.RuntimeException: Error initializing keystore
```

But this time I don't think it's in `web.xml`, so it might be easier to fix.

**Comment by fedde-s** [comment]
Found it—`applicationContext-security.xml` was configured to be edited by Maven at buildtime. It does contain a `propertyPlaceHolderConfigurer` to read the properties from the right `portal.properties` file (baked-in or `$PORTAL_HOME/`) itself though, so I that can safely be removed. I tested it with SAML using both buildtime and runtime configuration.

**Comment by Sjoerd-van-Hagen** [comment]
@n1zea144 hi Ben, can we try to move this one along?

File: docs/portal.properties-Reference.md [hunk 1]
```diff
@@ -73,11 +73,17 @@ googleplus.consumer.secret=2jCfg4SPWdGfXF44WC588dK
 ```
 (note: these are just examples, you need to get your own) You will also need to go to "Google+ API" and click Enable button. In case of problems make sure to enable DEBUG logging for org.springframework.social and org.springframework.security.web.authentication.
 
+To activate password authentication,
```
Comment (sheridancbio) [inline docs/portal.properties-Reference.md]: It seems like there should be a parallel document "Authenticating-and-Authorizing-Users-via-Googleplus.md" rather than showing that method here. Instead I would have put references here to the three/four/five? available guides on how to configure for different authentication methods.
But that change probably belongs in a different PR

**Comment by inodb** [comment]
i added one commit such that heroku is public by default

**Comment by sheridancbio** [comment]
This code has been built into the latest importers at mskcc and has not caused problems. Merging PR now.
